File: src/java/magmac/app/compile/error/CompileResultCollector.java
package magmac.app.compile.error;

import magmac.api.iter.collect.Collector;
import magmac.api.result.Ok;

public record CompileResultCollector<T, C>(Collector<T, C> joiner
) implements Collector<CompileResult<T>, CompileResult<C>> {
    @Override
    public CompileResult<C> createInitial() {
        return CompileResults.fromResult(new Ok<>(this.joiner.createInitial()));
    }

    @Override
    public CompileResult<C> fold(CompileResult<C> maybeCurrent, CompileResult<T> maybeElement) {
        return maybeCurrent.flatMapValue((C currentValue) -> maybeElement.mapValue((T element) -> this.joiner.fold(currentValue, element)));
    }
}

File: src/web/magmac/app/compile/error/CompileResultCollector.ts
import { Collector } from "../../../../magmac/api/iter/collect/Collector";
import { Ok } from "../../../../magmac/api/result/Ok";
export class CompileResultCollector<T,  C> {
	createInitial() : CompileResult<C> {return 0.fromResult( new Ok<>( 0.joiner.createInitial( )));;}
	fold(maybeCurrent : CompileResult<C>, maybeElement : CompileResult<T>) : CompileResult<C> {return 0.flatMapValue( 0);;}
}
